### ISSUE TITLE
Panic on log rotation when running `docker logs -f xxxx`

### ISSUE THREAD

**Issue by mountkin** [docker/docker#20327]
Reproduce:
```
docker run -d --name=test --log-opt max-size=500 --log-opt max-file=5 busybox sh -c "sleep 10;yes X|head -c 200"
docker logs -f test
```

Daemon log:
```
DEBU[23526] GET /v1.22/containers/logs/logs?follow=1&stderr=1&stdout=1&tail=all 
DEBU[23526] logs: begin stream                           
DEBU[23526] waiting for events                            logger=json-file
DEBU[23527] waiting for events                            logger=json-file
WARN[23527] falling back to file poller                   logger=json-file
ERRO[23527] error watching log file for modifications: stat /data1/docker/containers/f09b6bb01578d11ee96393cf9d99fb5ba32d54a941abf1520c1afea953980646/f09b6bb01578d11ee96393cf9d99fb5ba32d54a941abf1520c1afea953980646-json.log: no such file or directory 
DEBU[23527] waiting for events                            logger=json-file
WARN[23527] falling back to file poller                   logger=json-file
DEBU[23527] waiting for events                            logger=json-file
WARN[23527] falling back to file poller                   logger=json-file
DEBU[23527] waiting for events                            logger=json-file
ERRO[23527] Error streaming logs: stat /data1/docker/containers/f09b6bb01578d11ee96393cf9d99fb5ba32d54a941abf1520c1afea953980646/f09b6bb01578d11ee96393cf9d99fb5ba32d54a941abf1520c1afea953980646-json.log: no such file or directory 
WARN[23527] falling back to file poller                   logger=json-file
DEBU[23527] waiting for events                            logger=json-file
WARN[23527] falling back to file poller                   logger=json-file
DEBU[23527] waiting for events                            logger=json-file
WARN[23527] falling back to file poller                   logger=json-file
DEBU[23527] waiting for events                            logger=json-file
WARN[23527] falling back to file poller                   logger=json-file
DEBU[23527] waiting for events                            logger=json-file
WARN[23527] falling back to file poller                   logger=json-file
DEBU[23527] waiting for events                            logger=json-file
WARN[23527] falling back to file poller                   logger=json-file
ERRO[23527] error watching log file for modifications: open /data1/docker/containers/f09b6bb01578d11ee96393cf9d99fb5ba32d54a941abf1520c1afea953980646/f09b6bb01578d11ee96393cf9d99fb5ba32d54a941abf1520c1afea953980646-json.log: no such file or directory 
DEBU[23527] waiting for events                            logger=json-file
WARN[23527] falling back to file poller                   logger=json-file
ERRO[23527] error watching log file for modifications: open /data1/docker/containers/f09b6bb01578d11ee96393cf9d99fb5ba32d54a941abf1520c1afea953980646/f09b6bb01578d11ee96393cf9d99fb5ba32d54a941abf1520c1afea953980646-json.log: no such file or directory 
DEBU[23527] watch for /data1/docker/containers/f09b6bb01578d11ee96393cf9d99fb5ba32d54a941abf1520c1afea953980646/f09b6bb01578d11ee96393cf9d99fb5ba32d54a941abf1520c1afea953980646-json.log closed 
panic: send on closed channel

goroutine 980 [running]:
github.com/docker/docker/pkg/filenotify.(*filePoller).sendEvent(0xc82145a000, 0xc82179652a, 0x49, 0x2, 0xc8215780c0, 0x0, 0x0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/pkg/filenotify/poller.go:128 +0x164
github.com/docker/docker/pkg/filenotify.(*filePoller).watch(0xc82145a000, 0xc8207f54a0, 0x7fd26a65b648, 0xc8214d0000, 0xc8215780c0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/pkg/filenotify/poller.go:198 +0x676
created by github.com/docker/docker/pkg/filenotify.(*filePoller).Add
	/usr/src/docker/.gopath/src/github.com/docker/docker/pkg/filenotify/poller.go:69 +0x37e

goroutine 1 [chan receive, 392 minutes]:
main.(*DaemonCli).CmdDaemon(0xc8203ad760, 0xc82000a0c0, 0x8, 0x8, 0x0, 0x0)
	/usr/src/docker/docker/daemon.go:305 +0x20a7
reflect.callMethod(0xc8206e0de0, 0xc82093fc78)
	/usr/local/go/src/reflect/value.go:628 +0x1fc
reflect.methodValueCall(0xc82000a0c0, 0x8, 0x8, 0x1, 0xc8206e0de0, 0x0, 0x0, 0xc8206e0de0, 0x0, 0x479f34, ...)
	/usr/local/go/src/reflect/asm_amd64.s:29 +0x36
github.com/docker/docker/cli.(*Cli).Run(0xc8206e0d50, 0xc82000a0b0, 0x9, 0x9, 0x0, 0x0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/cli/cli.go:89 +0x383
main.main()
	/usr/src/docker/docker/docker.go:63 +0x43c

goroutine 17 [syscall, 392 minutes, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1721 +0x1
```

docker version:
```
Client:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Thu Feb 11 19:27:08 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Thu Feb 11 19:27:08 2016
 OS/Arch:      linux/amd64
```

docker info:
```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 232
Server Version: 1.10.1
Storage Driver: aufs
 Root Dir: /data1/docker/aufs
 Backing Filesystem: extfs
 Dirs: 242
 Dirperm1 Supported: false
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: null host bridge
Kernel Version: 3.13.0-76-generic
Operating System: Ubuntu 14.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 7.703 GiB
Name: sky
ID: HJRN:G4UP:MRFS:DPGA:UETY:4533:2KPV:EIXV:R5T7:N5IU:YQWU:TKK2
Debug mode (server): true
 File Descriptors: 23
 Goroutines: 59
 System Time: 2016-02-15T19:32:01.93920522+08:00
 EventsListeners: 0
 Init SHA1: e1042dbb0bcf49bb9da188176d9a5063cdb92a01
 Init Path: /usr/lib/docker/dockerinit
 Docker Root Dir: /data1/docker
WARNING: No swap limit support
```

**Comment by mountkin** [comment]
#19498 is trying to remove the `filenotify` package and use `pubsub` to follow logs.
#19498 solved the panic problem but it seems that the log is not correctly followed on rotation (with "bad file descriptor" error when rotation happens).

**Comment by thaJeztah** [comment]
Do you know if this is a regression in 1.10, or was this already the case in 1.9?

**Comment by mountkin** [comment]
It's probably not a regression but I can't confirm it until tomorrow morning(GMT +8) because I haven't got a computer around :innocent: 
The daemon will fail to start after the panic. The error message is the same as https://github.com/docker/docker/issues/17083 
I have to delete the local kv file to start the daemon.

**Comment by thaJeztah** [comment]
Thanks @mountkin, keep us posted; if it is a regression, I'm tempted to add it to the 1.10.2 milestone

**Comment by thaJeztah** [comment]
hm weird, I wasn't able to reproduce this using your steps, but I didn't get `falling back to file poller`, so perhaps that's a difference

**Comment by sfarges** [comment]
I have the same issue. for info I reproduced with @mountkin steps after 10 attempts of :  
```
docker rm test
docker run -d --name=test --log-opt max-size=500 --log-opt max-file=5 busybox sh -c "sleep 2;yes X|head -c 200"
docker logs -f test
sudo cat /var/log/syslog | grep poller
```

I have exactly same docker version. But the host is on ubuntu 15.10:
````
docker info
Containers: 9
 Running: 7
 Paused: 0
 Stopped: 2
Images: 43
Server Version: 1.10.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 106
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 4.2.0-27-generic
Operating System: Ubuntu 15.10
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 1.907 GiB
Name: xxxxxx
ID: P7OB:UAM6:U56Z:L4EF:TZV7:ZD22:FSFG:KBCG:2PV4:NYHZ:GC4L:QFSH
WARNING: No swap limit support
```

**Comment by mountkin** [comment]
@thaJeztah The panic was introduced by #17417 and it's in 1.10.x only.
Now that #19498  is merged and the panic is fixed in master. I think this issue can be closed.

**Comment by mountkin** [comment]
@thaJeztah You have to run `docker logs -f test` before the container stopped.
For convenience you can reproduce the problem by 
```
#!/bin/bash
set -e
for ((i=0;i<100;i++));do 
  docker rm test >/dev/null 2>&1 || true
  docker run -d --name=test --log-opt max-size=500 \
    --log-opt max-file=5 busybox sh -c "sleep 2;yes X|head -c 200"
  docker logs -f test
done
```

**Comment by thaJeztah** [comment]
Thanks! If its in 1.10 only, this should be on the 1.10.2 milestone as a regression

**Comment by tiborvass** [comment]
I'm reopening since we don't feel comfortable taking mountkin's fix for 1.10.2.

**Comment by tiborvass** [comment]
Fixed by #20522
